File: SwiftGameOfLife/GameOfLife/Cell.swift
//
//  Cell.swift
//  GameOfLife
//
//  Created by Eirik Vale Aase on 02/04/2022.
//

final class Cell {
    enum State {
        case dead
        case alive
    }

    var coordinate: Coordinate
    var representation: String = "_"
    var nextState: State?

    init(coordinate: Coordinate) {
        self.coordinate = coordinate
    }

    var state: State {
        switch representation {
        case "_": return .dead
        case "x": return .alive
        default: fatalError("Unsupported representation.")
        }
    }

    func kill() {
        representation = "_"
    }

    func birth() {
        representation = "x"
    }

    func handleNextState() {
        switch nextState {
        case .alive?:
            representation = "x"
        case .dead?:
            representation = "_"
        default:
            break
        }
    }
}

File: PythonGameOfLife/Cell.py
from CellState import CellState
from Coordinate import Coordinate


class Cell:
    def __init__(self, coordinate: Coordinate):
        self.coordinate = coordinate
        self.representation = "_"
        # TODO: Look into modelling this as an optional property.
        self.nextState = CellState.DEAD

    def get_state(self):
        if self.representation == "_":
            return CellState.DEAD
        elif self.representation == "x":
            return CellState.ALIVE
        else:
            raise ValueError("Unsupported representation.")

    def kill(self):
        self.representation = "_"

    def birth(self):
        self.representation = "x"

    def handle_next_state(self):
        if self.nextState is CellState.ALIVE:
            self.birth()
        elif self.nextState is CellState.DEAD:
            self.kill()
        else:
            raise ValueError("Unsupported state.")

    def __repr__(self):
        return self.representation
